### PR TITLE
Add `Memory<byte>` support to WebSockets

### PR DESCRIPTION
**Pull request by stephentoub**
This is split into several commits as there are multiple related but distinct changes:
1. First, I moved ManagedWebSocket from being based on `ArraySegment<byte>` internally to being based on `Memory<byte>` internally.  That then allows it to be used with either `ArraySegment<byte>` or `Memory<Byte>`, as the former is convertible to the latter.
2. Then I added the new `CreateFromStream` method to WebSocket; this requires the `Memory<byte>` support in ManagedWebSocket because `CreateFromStream`'s scratch buffer is a `Memory<byte>`.
3. Then I used that new `CreateFromStream` method in both System.Net.WebSockets.Client and System.Net.HttpListener; ManagedWebSocket.cs now is only built into System.Net.WebSockets rather than into all three assemblies.
4. Then I added the new `Memory<byte>`-based Send/ReceiveAsync virtuals to WebSocket.
5. Then I overrode these on ManagedWebSocket, providing decent implementations for them.
6. And finally I overrode the `Memory<byte>`-based `ReceiveAsync` on WinHttpWebSocket.  I didn't bother with `SendAsync` for now as the base implementation should be reasonable for common cases, and if we want to optimize this implementation, there are other things that can and should be done, too.

cc: @geoffkizer, @davidsh, @anurse, @benaadams, @Priya91 
Fixes https://github.com/dotnet/corefx/issues/22610
Contributes to https://github.com/dotnet/corefx/issues/21537 (we still need to expose it through an OOB)

### CODE REVIEW

**Comment by stephentoub** [comment]
@dotnet-bot test Windows x86 Release Build please (https://github.com/dotnet/corefx/issues/23341)

**Comment by stephentoub** [comment]
@dotnet-bot test Linux x64 Release Build please (https://github.com/dotnet/corefx/issues/24775)

**Comment by stephentoub** [comment]
@dotnet-bot test Linux x64 Release Build please (#24775)

**Comment by stephentoub** [comment]
@dotnet-bot test Linux x64 Release Build please (#24775)

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please
@dotnet-bot test Outerloop Linux x64 Debug Build please

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Linux x64 Debug Build please (https://github.com/dotnet/corefx/issues/24875, https://github.com/dotnet/corefx/issues/23072, https://github.com/dotnet/corefx/issues/17681)

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Linux x64 Debug Build please (https://github.com/dotnet/corefx/issues/23535)

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Linux x64 Debug Build please (https://github.com/dotnet/corefx/issues/23799)

**Comment by stephentoub** [comment]
@dotnet/dnceng, the Outerloop Linux x64 Debug Build leg failed, saying that the System.Collections.Concurrent tests were on fire, but the log shows them completing successfully, and then this:
https://mc.dot.net/#/user/stephentoub/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/98d53f272c58b4df9ce997d816d853ddff43e7d8/workItem/System.Collections.Concurrent.Tests/wilogs
```
2017-10-26 00:47:16,950: INFO: proc(54): run_and_log_output: Output: === TEST EXECUTION SUMMARY ===
2017-10-26 00:47:16,956: INFO: proc(54): run_and_log_output: Output:    System.Collections.Concurrent.Tests  Total: 2087, Errors: 0, Failed: 0, Skipped: 0, Time: 36.287s
2017-10-26 00:47:17,037: INFO: proc(54): run_and_log_output: Output: ~/dotnetbuild/work/85769064-c8e1-4faf-87b1-2fa16c112032/Work/12b78a72-fe2e-4740-a695-a2e4087f9704/Unzip
2017-10-26 00:47:17,038: INFO: proc(54): run_and_log_output: Output: Finished running tests. End time=00:47:17. Return value was 0
2017-10-26 00:47:17,040: INFO: proc(58): run_and_log_output: Exit Code: 0
2017-10-26 00:47:17,040: INFO: scriptrunner(62): _main: Uploading results from /home/helixbot/dotnetbuild/work/85769064-c8e1-4faf-87b1-2fa16c112032/Work/12b78a72-fe2e-4740-a695-a2e4087f9704/Unzip/testResults.xml
2017-10-26 00:47:17,040: INFO: azure_storage(205): _upload: Uploading single blob: 'testResults.xml'
2017-10-26 00:47:17,046: WARNING: saferequests(70): request_with_retry: Request failed with exception: ConnectionError(ProtocolError('Connection aborted.', gaierror(-2, 'Name or service not known')),)
2017-10-26 00:47:17,047: WARNING: saferequests(70): request_with_retry: Request failed with exception: ConnectionError(ProtocolError('Connection aborted.', gaierror(-2, 'Name or service not known')),)
2017-10-26 00:47:17,047: INFO: saferequests(64): request_with_retry: Waiting 1.36423338538 seconds for retry 2
2017-10-26 00:47:18,414: WARNING: saferequests(70): request_with_retry: Request failed with exception: ConnectionError(ProtocolError('Connection aborted.', gaierror(-2, 'Name or service not known')),)
2017-10-26 00:47:18,414: INFO: saferequests(64): request_with_retry: Waiting 1.35740099494 seconds for retry 3
2017-10-26 00:47:19,774: WARNING: saferequests(70): request_with_retry: Request failed with exception: ConnectionError(ProtocolError('Connection aborted.', gaierror(-2, 'Name or service not known')),)
2017-10-26 00:47:19,775: INFO: saferequests(64): request_with_retry: Waiting 4.16389688975 seconds for retry 4
2017-10-26 00:47:23,944: WARNING: saferequests(70): request_with_retry: Request failed with exception: ConnectionError(ProtocolError('Connection aborted.', gaierror(-2, 'Name or service not known')),)
2017-10-26 00:47:23,944: ERROR: saferequests(104): request_with_retry: Retry count 5 exceeded, re-raising last error
Traceback (most recent call last):
  File "/home/helixbot/dotnetbuild/work/85769064-c8e1-4faf-87b1-2fa16c112032/Payload/RunnerScripts/scriptrunner/scriptrunner.py", line 96, in <module>
    sys.exit(main())
  File "/home/helixbot/dotnetbuild/work/85769064-c8e1-4faf-87b1-2fa16c112032/Payload/RunnerScripts/scriptrunner/scriptrunner.py", line 92, in main
    return command_main(_main, ['script=', 'args='], args)
  File "/home/helixbot/dotnetbuild/scripts/helix/cmdline.py", line 47, in command_main
    return main_method(settings, optlist, args)
  File "/home/helixbot/dotnetbuild/work/85769064-c8e1-4faf-87b1-2fa16c112032/Payload/RunnerScripts/scriptrunner/scriptrunner.py", line 74, in _main
    result_url = test_executor.upload_file_to_storage(results_location, settings)
  File "/home/helixbot/dotnetbuild/scripts/helix_test_execution.py", line 37, in upload_file_to_storage
    url = fc.upload(file_path, os.path.basename(file_path))
  File "/home/helixbot/dotnetbuild/scripts/helix/azure_storage.py", line 77, in upload
    self._upload(f, name)
  File "/home/helixbot/dotnetbuild/scripts/helix/azure_storage.py", line 208, in _upload
    retry_status_ranges=helix.saferequests.RETRY_RANGE_ALL_ERROR
  File "/home/helixbot/dotnetbuild/scripts/helix/saferequests.py", line 105, in request_with_retry
    raise last_error
requests.exceptions.ConnectionError: ('Connection aborted.', gaierror(-2, 'Name or service not known'))
2017-10-26 00:47:23,945: ERROR: logs(87): _log_uncaught_exception: Unhandled error: ('Connection aborted.', gaierror(-2, 'Name or service not known'))
['  File "/home/helixbot/dotnetbuild/work/85769064-c8e1-4faf-87b1-2fa16c112032/Payload/RunnerScripts/scriptrunner/scriptrunner.py", line 96, in <module>\n    sys.exit(main())\n', '  File "/home/helixbot/dotnetbuild/work/85769064-c8e1-4faf-87b1-2fa16c112032/Payload/RunnerScripts/scriptrunner/scriptrunner.py", line 92, in main\n    return command_main(_main, [\'script=\', \'args=\'], args)\n', '  File "/home/helixbot/dotnetbuild/scripts/helix/cmdline.py", line 47, in command_main\n    return main_method(settings, optlist, args)\n', '  File "/home/helixbot/dotnetbuild/work/85769064-c8e1-4faf-87b1-2fa16c112032/Payload/RunnerScripts/scriptrunner/scriptrunner.py", line 74, in _main\n    result_url = test_executor.upload_file_to_storage(results_location, settings)\n', '  File "/home/helixbot/dotnetbuild/scripts/helix_test_execution.py", line 37, in upload_file_to_storage\n    url = fc.upload(file_path, os.path.basename(file_path))\n', '  File "/home/helixbot/dotnetbuild/scripts/helix/azure_storage.py", line 77, in upload\n    self._upload(f, name)\n', '  File "/home/helixbot/dotnetbuild/scripts/helix/azure_storage.py", line 208, in _upload\n    retry_status_ranges=helix.saferequests.RETRY_RANGE_ALL_ERROR\n', '  File "/home/helixbot/dotnetbuild/scripts/helix/saferequests.py", line 105, in request_with_retry\n    raise last_error\n']
2017-10-26 00:47:23,945: INFO: azure_storage(205): _upload: Uploading single blob: 'acb05482c83040cf9f5900714bd006da.log'
2017-10-26 00:47:23,946: WARNING: saferequests(70): request_with_retry: Request failed with exception: ConnectionError(ProtocolError('Connection aborted.', gaierror(-2, 'Name or service not known')),)
2017-10-26 00:47:23,947: WARNING: saferequests(70): request_with_retry: Request failed with exception: ConnectionError(ProtocolError('Connection aborted.', gaierror(-2, 'Name or service not known')),)
2017-10-26 00:47:23,947: INFO: saferequests(64): request_with_retry: Waiting 1.21858558879 seconds for retry 2
2017-10-26 00:47:25,295: INFO: saferequests(90): request_with_retry: Response complete with status code '201'
2017-10-26 00:47:25,295: INFO: event(43): send: Sending event type Logs
2017-10-26 00:47:25,595: INFO: saferequests(90): request_with_retry: Response complete with status code '201'
'module' object has no attribute '__file__'
```

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Linux x64 Debug Build please

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Linux x64 Debug Build please (https://github.com/dotnet/corefx/issues/23535)

**Comment by mmitche** [comment]
@stephentoub @dotnet/dnceng https://github.com/dotnet/core-eng/issues/1958

File: src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs [hunk 1]
```diff
@@ -313,6 +321,29 @@ public override Task<WebSocketReceiveResult> ReceiveAsync(ArraySegment<byte> buf
...
+                WebSocketValidate.ThrowIfInvalidState(_state, _disposed, s_validReceiveStates);
+
+                Debug.Assert(!Monitor.IsEntered(StateUpdateLock), $"{nameof(StateUpdateLock)} must never be held when acquiring {nameof(ReceiveAsyncLock)}");
+                lock (ReceiveAsyncLock) // synchronize with receives in CloseAsync
+                {
+                    ThrowIfOperationInProgress(_lastReceiveAsync);
+                    ValueTask<ValueWebSocketReceiveResult> t = ReceiveAsyncPrivate<ValueWebSocketReceiveResultGetter, ValueWebSocketReceiveResult>(buffer, cancellationToken);
+                    _lastReceiveAsync =
+                        t.IsCompletedSuccessfully ? (t.Result.MessageType == WebSocketMessageType.Close ? s_cachedCloseTask : Task.CompletedTask) :
+                        t.AsTask();
```
Comment (benaadams) [inline src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs]: This bit is pretty neat `_lastReceiveAsync` 😄

File: src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs [hunk 2]
```diff
@@ -705,24 +743,25 @@ private static int WriteHeader(MessageOpcode opcode, byte[] sendBuffer, ArraySeg
...
                     }
-                    Buffer.BlockCopy(_receiveBuffer, _receiveBufferOffset, payloadBuffer.Array, payloadBuffer.Offset, bytesToCopy);
+                    _receiveBuffer.Span.Slice(_receiveBufferOffset, bytesToCopy).CopyTo(payloadBuffer.Span.Slice(0, bytesToCopy));
                     ConsumeFromBuffer(bytesToCopy);
                     header.PayloadLength -= bytesToCopy;
 
                     // If this a text message, validate that it contains valid UTF8.
                     if (header.Opcode == MessageOpcode.Text &&
-                        !TryValidateUtf8(new ArraySegment<byte>(payloadBuffer.Array, payloadBuffer.Offset, bytesToCopy), header.Fin, _utf8TextState))
+                        !TryValidateUtf8(payloadBuffer.Span.Slice(0, bytesToCopy), header.Fin, _utf8TextState))
```
Comment (benaadams) [inline src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs]: Future candidate for `Utf8Utility.IsWellFormedUtf8String`? https://github.com/dotnet/corefxlab/pull/1881 or alternative routine for benchmarking against. 

/cc @GrabYourPitchforks

**Comment by benaadams** [comment]
LGTM

File: src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs (hunk 2 above)
Comment (GrabYourPitchforks) [reply src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs]: Thanks for pointing out this use case! Solidifies my belief that I need to add a streaming API to the validator.

Comment (GrabYourPitchforks) [reply src/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs]: https://github.com/dotnet/corefxlab/pull/1918 should provide a centralized drop-in replacement for this logic.

**Pull request merged**